MTA-6439: grammar issues (#352)
migtools/mta-documentation: docs/topics/mta-cli/proc_generate_deployment_assets_values_file.adoc

= Generating deployment manifests from a discovery manifest and values.yaml

[role="_abstract"]
When you run the `generate helm` command, you provide a `Helm` chart by using `--chart-dir`. The `values.yaml` in that chart is used to capture keys that are not present in the discovery manifest. You can also override specific values by using the `--set` flag when you run the `generate` command. 

> To use the `helm` templates to generate assets, save the Kubernetes `YAML` templates, for example, `ConfigMap` or `Deployment`, in the `templates` directory present in the root directory of the `Helm` chart. You must save non-Kubernetes templates, for example, `Dockerfile`, in the `files/konveyor` path in the `Helm` chart root directory and save the `values.yaml` file at the root directory of the `Helm` chart.

Reviewer: @Pkylas007, your reply for this block somehow ended as a part of my comment which now looks like I'm talking to myself again :-D 

Here are my thoughts: Since this is something that the user should've done at some point, I'd add it as a part of Prerequisites, for example, like that (also, some formatting issues fixed):

```
* To use Helm templates to generate assets, you saved your files in the Helm chart root directory as follows:
** You saved Kubernetes YAML templates, for example, `ConfigMap` or `Deployment`, in the `templates` directory.
** You saved non-Kubernetes templates, for example, `Dockerfile`, in the `files/konveyor` path.
** You saved the `values.yaml` file in the root directory.
```

> * `Helm CLI` is installed.

Author: ```suggestion
* `Helm` command-line interface (CLI) is installed.
```

> * `Helm` command-line interface (CLI) is installed.

Reviewer: If we decide on having another prerequisite (see above), I suggest unifying the style:
```suggestion
* You installed the `Helm` command-line interface (CLI).
```

.Procedure

. Define the application name, a database host, and a Redis database URL in a `values.yaml` file:
+

[source, yaml]
----
manifest:
  name: my-app
  database:
    enabled: true
    host: prod-postgres.company.com
  redis:
    url: redis://prod-redis:6379
----
+

> 

Reviewer: Noticed there are many empty lines where they-re not needed
```suggestion
```

[NOTE]
====
In this example, the discovery manifest does not contain the fields in the `values.yaml` file.
====

. Create a `Dockerfile` template that uses environment variables defined in the `values.yaml` file:
+

> 

Reviewer: ```suggestion
```

= Generating deployment manifests from a discovery manifest and values.yaml

[role="_abstract"]
Use a Helm chart and a `values.yaml` file to generate deployment manifests that include configuration keys missing from a discovery manifest. This process allows you to capture additional metadata or use the `--set` flag to override values and customize your generated assets.

.Prerequisites

> * `Helm` command-line interface (CLI) is installed.

Author: ```suggestion
* * You installed the `Helm` command-line interface (CLI).
* To use Helm templates to generate assets, you saved your files in the Helm chart root directory as follows:
** You saved Kubernetes YAML templates, for example, `ConfigMap` or `Deployment`, in the `templates` directory.
** You saved non-Kubernetes templates, for example, `Dockerfile`, in the `files/konveyor` path.
** You saved the `values.yaml` file in the root directory.
```

= Generating deployment manifests from a discovery manifest and values.yaml

[role="_abstract"]
Use a Helm chart and a `values.yaml` file to generate deployment manifests that include configuration keys missing from a discovery manifest. This process allows you to capture additional metadata or use the `--set` flag to override values and customize your generated assets.

.Prerequisites

> * * You installed the `Helm` command-line interface (CLI).

Author: ```suggestion
* You installed the `Helm` command-line interface (CLI).
```

ENV REDIS_URL={{ .Values.manifest.redis.url }}
{{- end }}
----
. Generate deployment assets, such as a `Dockerfile`, that use the variables defined in the `values.yaml` file:
+
[source, terminal]

> ----

Reviewer: The command doesn't render correctly in the block: https://docs.redhat.com/en/documentation/migration_toolkit_for_applications/8.1/html/using_the_migration_toolkit_for_applications_command-line_interface/generating-assets_cli-guide#values-deployment_generating-assets. It renders asterisks as plain text. Probably, a problem of the code block as a whole. We usually use [subs="+quotes"] there.

Reviewer 2: Please shall we remove the asterisks? Does it serve any purpose in a published page?

Reviewer: No, we cannot. Asterisks make a command bold. And we're explicitly instructed to use this format for commands the user must enter.

  *--input _<path_to_discovery-manifest>_* \
  *--set manifest.name=my-new-app* \
  *--output-dir _<location_of_deployment_manifests>_*
----
+
where:
`_<path_to_helm_chart>_`:: Specifies the directory containing the Helm chart.
`_<path_to_discovery_manifest>_`:: Specifies the location of the discovery manifest file.
`_<location_of_deployment_manifests>_`:: Specifies the output directory for the generated manifests.

+
[NOTE]
====
* The `--set` flag value has priority. If you specify a variable by using the `--set` flag, that value overrides the discrete variable defined in the `values.yaml` file or the discovery manifest.

* The discovery manifest configuration has the second priority. If you do not use the `--set` flag, {ProductShortName} uses discovery manifest values as the input for generation.

* If the `values.yaml` file is not empty, the generated deployment manifests use those values as input. If a key exists in both the discovery manifest and the values.yaml file, {ProductShortName} uses the value from the discovery manifest.
====

.Verification

* Verify that the rendered `Dockerfile` includes the fields from the `values.yaml` file:
+

> [source, terminal]

Author: ```suggestion
 [subs="+quotes"]
```

ENV REDIS_URL={{ .Values.manifest.redis.url }}
{{- end }}
----
. Generate deployment assets, such as a `Dockerfile`, that use the variables defined in the `values.yaml` file:
+
[subs="+quotes"]
----
$ *mta-cli generate helm --chart-dir _<path_to_helm_chart>_* \

>   *--input _<path_to_discovery-manifest>_* \

Reviewer: ```suggestion
  *--input _<path_to_discovery_manifest>_* \
```

  *--set manifest.name=my-new-app* \
  *--output-dir _<location_of_deployment_manifests>_*
----
+

> where: 

Reviewer: ```suggestion
where:

```

`_<path_to_helm_chart>_`: Specifies the directory containing the Helm chart. 
`_<path_to_discovery_manifest>_`: Specifies the location of the discovery manifest file. 

> `_<location_of_deployment_manifests>_`: Specifies the output directory for the generated manifests. 

Reviewer: Otherwise, it won't render correctly
```suggestion
`_<path_to_helm_chart>_`:: Specifies the directory containing the Helm chart. 
`_<path_to_discovery_manifest>_`:: Specifies the location of the discovery manifest file. 
`_<location_of_deployment_manifests>_`:: Specifies the output directory for the generated manifests. 
```

> `_<location_of_deployment_manifests>_`:: Specifies the output directory for the generated manifests. 

Reviewer: Actually, a great catch! This should be consistent. We use underscores for such values, so _\<path_to_discovery_manifest>_ is the correct option.

> `_<location_of_deployment_manifests>_`:: Specifies the output directory for the generated manifests. 

Author: @mpershina -this look weird when it renders, please can you have a look-see:

<img width="1009" height="336" alt="image" src="https://github.com/user-attachments/assets/58abd8fe-ea33-4460-8d92-9faac35a9c80" />

Is it ok to change to

<img width="1877" height="453" alt="image" src="https://github.com/user-attachments/assets/4a718748-f42a-4a04-9a52-df7cb3d55305" />

Thanks

Reviewer: Oh, I think I know where the issue might be :-D 
That's why we always need a preview. So, we basically need to have an empty line after `where`. I added my suggestion above

+
The variable priority hierarchy for generating deployment manifests is as follows:

* **Priority 1: The `--set` command flag**: Any value specified with this flag has the highest priority and overrides variables defined in either the `values.yaml` file or the discovery manifest.
* **Priority 2: The discovery manifest**: If you do not use the `--set` flag, the discovery manifest configuration is used as the primary input for generation.

> * **Priority 3: The `values.yaml` file**: If this file is not empty, it provides the input values; however, if a key exists in both the discovery manifest and the `values.yaml` file, the value from the discovery manifest takes precedence.

Reviewer: Just a nit :)
```suggestion
* **Priority 3: The `values.yaml` file**: If this file is not empty, it provides the input values. However, if a key exists in both the discovery manifest and the `values.yaml` file, the value from the discovery manifest takes precedence.
```

ENV REDIS_URL={{ .Values.manifest.redis.url }}
{{- end }}
----
. Generate deployment assets, such as a `Dockerfile`, that use the variables defined in the `values.yaml` file:
+
[subs="+quotes"]
----
$ *mta-cli generate helm --chart-dir _<path_to_helm_chart>_* \
  *--input _<path_to_discovery_manifest>_* \
  *--set manifest.name=my-new-app* \
  *--output-dir _<location_of_deployment_manifests>_*
----
+
where:

`_<path_to_helm_chart>_`:: Specifies the directory containing the Helm chart. 
`_<path_to_discovery_manifest>_`:: Specifies the location of the discovery manifest file. 
`_<location_of_deployment_manifests>_`:: Specifies the output directory for the generated manifests. 
+
The variable priority hierarchy for generating deployment manifests is as follows:

> * **Priority 1: The `--set` command flag**: Any value specified with this flag has the highest priority and overrides variables defined in either the `values.yaml` file or the discovery manifest.

Reviewer 2: ```suggestion
* **(Priority 1) The `--set` command flag**: Any value specified with this flag has the highest priority and overrides variables defined in either the `values.yaml` file or the discovery manifest.
```
Please shall we remove too many colons in a line here? The format looks a little weird, WDYT?

ENV REDIS_URL={{ .Values.manifest.redis.url }}
{{- end }}
----
. Generate deployment assets, such as a `Dockerfile`, that use the variables defined in the `values.yaml` file:
+

> [source, terminal]

Author: ```suggestion
[subs="+quotes"]
```

  *--output-dir _<location_of_deployment_manifests>_*
----
+
* `<path_to_helm_chart>`: Specifies the directory containing the Helm chart.
* `<path_to_discovery_manifest>`: Specifies the location of the discovery manifest file.

> * `<location_of_deployment_manifests>`: Specifies the output directory for the generated manifests.

Reviewer: Wrong formatting for the user-replaced values (https://redhat-documentation.github.io/supplementary-style-guide/#user-replaced-values).

Also, per our current guidelines (checked with Notebook), apparently, it's ok to use the `where` format, which I don't really like, TBH. I will check with Ingrid whether these is something they plan to change
```suggestion
where:
`_<path_to_helm_chart>_`:: Specifies the directory containing the Helm chart.
`_<path_to_discovery_manifest>_`:: Specifies the location of the discovery manifest file.
`_<location_of_deployment_manifests>_`:: Specifies the output directory for the generated manifests.
```

Reviewer: A bulleted list is also an option, but in our instructions, it goes without any intro sentence, which I don't really like. I'd prefer the description list format in this case

> * `<location_of_deployment_manifests>`:: Specifies the output directory for the generated manifests.

Author: ```suggestion
* `<path_to_helm_chart>`: Specifies the directory containing the Helm chart.
* `<path_to_discovery_manifest>`: Specifies the location of the discovery manifest file.
* `<location_of_deployment_manifests>`: Specifies the output directory for the generated manifests.
```

+
[NOTE]
====
* The `--set` flag value has priority. If you specify a variable by using the `--set` flag, that value overrides the discrete variable defined in the `values.yaml` file or the discovery manifest.

* The discovery manifest configuration has the second priority. If you do not use the `--set` flag, {ProductShortName} uses discovery manifest values as the input for generation.

> * If the `values.yaml` file is not empty, the generated deployment manifests use those values as input. If a key exists in both the discovery manifest and the values.yaml file, {ProductShortName} uses the value from the discovery manifest.

Reviewer: Got some nice style feedback from Notebook here:

Areas for Improvement
1. Misuse of the NOTE Admonition The style guidelines state that a NOTE should be used for "additional guidance or advice that improves product configuration, performance, or supportability." Your list explains the core order of precedence for configuration values. Because this is standard, essential product behavior rather than an optional tip or aside, it should be moved out of the admonition and placed into the main body text. Additionally, the guidelines mandate that admonitions be "short and concise," making large lists inside them less ideal.
2. List Type Because your list describes a strict hierarchy (first priority, second priority, third priority), it should be formatted as an ordered (numbered) list rather than an unordered bulleted list. The guidelines explicitly state: "Not all numbered lists in documentation are procedures. You can also use numbered lists in any module type for non-procedural sequences."
3. Missing Monospace Formatting In the third bullet, the second instance of the file name values.yaml is missing its backticks. File names must consistently use monospace formatting (e.g., `values.yaml`).
4. Minimalism and Clarity Applying Red Hat's minimalism principles (eliminating fluff and improving scannability), the text can be tightened up to avoid repeating the exact conditions of what overrides what in every single bullet.

So, what about the following rewrite:
```suggestion
{ProductShortName} uses the following order of precedence to determine the input values for the generated deployment manifest:

* The `--set` flag: Values specified by using the `--set` flag have the highest priority. They override any variables defined in the discovery manifest or the `values.yaml` file.
* The discovery manifest configuration: Values in the discovery manifest configuration have the second priority. If you do not use the `--set` flag, {ProductShortName} uses discovery manifest values as the input for generation.
* The `values.yaml` file: If the `values.yaml` file is not empty, the generated deployment manifest uses its values as input. If a key exists in both the discovery manifest and the `values.yaml` file, {ProductShortName} uses the value from the discovery manifest.
```

Reviewer 2: Sorry but this would make it a numbered list, right? I think it must be an unordered list inside a NOTE or otherwise. It's not a set of procedure but important information for the user to know.

Reviewer: We can turn it in the bulleted list, although, it's not strictly prohibited to use ordered lists in admonitions.

Reviewer 2: Ay, no...there's no order of operations to be performed by the user here. Just priority info.

Reviewer: And this is why the ordered list, exactly because of the priority info. As I said, we can turn into a bulleted list, but it might break the logic a bit.

Reviewer 2: I think a bulleted list is a more suitable option here because priority info about the files when rendered is not equal to the order in which the user must add those files (procedural info).

====

.Verification

> * Verify that the rendered `Dockerfile` contains the fields from the `values.yaml` file:

Reviewer: Not sure if this is critical, but my Vale and other tools always complain about "contains"..
```suggestion
* Verify that the rendered `Dockerfile` includes the fields from the `values.yaml` file:
```

> +

Reviewer: Same here, asterisks render as plain text. What about trying [subs="+quotes"]?

Reviewer 2: [subs="+quotes"] doesn't render properly. The subs attribute shows in the published doc. That's why it was changed to this format.

Reviewer: That is really strange. I have this format across all of my procedures, and it renders as it should: https://docs.redhat.com/en/documentation/migration_toolkit_for_applications/8.1/html/using_the_migration_toolkit_for_applications_command-line_interface/analyzing-nonjava-applications_cli-guide#analyze-selected-provider_analyzing-nonjava-applications
